sklearn/model_selection/tests/test_search.py

```diff
@@ -641,6 +641,110 @@ def test_refit():
     clf.fit(X, y)
 
 
+def test_refit_callable():
+    """
+    Test refit=callable, which adds flexibility in identifying the
+    "best" estimator.
+    """
+    def refit_prec(cv_results):
+        """
+        Balance model complexity with cross-validated score.
+        Return the index of a model that has the least `mean_fit_time`,
+        but has test score within 1 standard deviation of the best
+        `mean_test_score`.
+        """
+        std_test_score = np.std(cv_results['mean_test_score'])
+        best_test_score = max(cv_results['mean_test_score'])
+        test_score_upper = np.minimum(best_test_score + std_test_score, 1)
+        test_score_lower = np.maximum(best_test_score - std_test_score, 0)
+        test_scores = {k: v for k, v in
+                       enumerate(cv_results['mean_test_score'])}
+        # Select models which have test scores within 1 standard deviation
+        # of the best 'mean_test_score'
+        candidates = dict(filter(lambda i: (i[1] >= test_score_lower
+                                 and i[1] <= test_score_upper),
+                                 test_scores.items()))
+        fit_time = {v: k for k, v in
+                    enumerate(cv_results['mean_fit_time'])}
+        fit_time_rank = sorted(fit_time)
+        for i in fit_time_rank:
+            if fit_time[i] in candidates:
+                # Find the index of a model that has the least
+                # 'mean_fit_time' while has a test score within
+                # 1 standard deviation of the best 'mean_test_score'
+                return fit_time[i]
+
+    X, y = make_classification(n_samples=100, n_features=4,
+                               random_state=42)
+    clf = GridSearchCV(LinearSVC(random_state=42), {'C': [0.01, 0.1, 1]},
+                       scoring="precision", refit=refit_prec)
+    clf.fit(X, y)
+    assert_equal(clf.best_index_, 1)
+
+
+def test_refit_callable_multi_metric():
+    """
+    Test refit=callable when using multi-metric scoring
+    For multi-metric evaluation, the name of refit callable function must
+    end with a scorer key(`_<scorer_name>`).
+    """
+    def refit_prec(cv_results):
```

Comment:
We should have a realistic example in `examples/model_selection/` rather than here.

As a simple example, I would consider using maximising score while minimising the number of selected features or PCA components.

Here we should merely be testing interface, and a dummy function (for instance, one that always chooses the lowest-score model) is sufficient / most appropriate, as it is then easy for us to be sure what correct behaviour is.

Reply:
@jnothman Would you say a dummy function like below is good enough to test our interface? 

```python
def refit_callable(cv_results):
    return cv_results['mean_test_score'].argmin()
```

Reply:
Yes, that looks good. I might add to that an assertion that all the keys we expect to be in results are in there.

Yes, I am indeed suggesting a second thing here. An example in `examples/model_selection` will hugely increase the visibility and practical usability of this feature. The example gallery is how we advise users how to use the features described in technical detail in the docstrings (and before StackOverflow has all the answers).

Reply:
Thanks! I'm adding a example from `examples/model_selection` for this feature in docstring.

Reply:
@jnothman is it appropriate to add one more example for `refit=callable` in the docstring under GridSearchCV class after this one?
 https://github.com/scikit-learn/scikit-learn/blob/3b5abf76597ce6aff76192869f92647c1b5259e7/sklearn/model_selection/_search.py#L931-L958

Reply:
I think a meaningful example is too large, and too much of a power-user feature, to be in the docstring.

Reply:
@jnothman It seems that we dont need to test our example under `examples` directory, right? ;)